File: hash_advanced/07_minimum_complete/python3/minimum_complete.py
# これはpaizaラーニングの「ハッシュメニュー応用編」-「最小完全ハッシュ関数を作ろう」に
# Python3でチャレンジしたコードです。
# https://paiza.jp/works/mondai/hash_advanced/hash_advanced__minimum_complete
# 
# 作成環境
# Ubuntu 22.04.5 LTS
# Python 3.10.12 (main, Nov  6 2024, 20:22:13) [GCC 11.4.0]

# ハッシュ関数
# parameters
#   s 文字列
# return sのハッシュ値
def hashFunction(s):
    string = 'abcdef'
    total = 0
    length = len(s)
    for i in range(length):
        index = string.find(s[length - i - 1])
        total += index * 6 ** i
    return total

# 最小完全ハッシュ関数を作る

# 文字列 s が与えられます。
s = input().strip()
# ハッシュ値を求める
hashValue = hashFunction(s)
# 結果出力
print(hashValue)

File: hash_advanced/07_minimum_complete/swift/minimum_complete.swift
/*
これはpaizaラーニングの「ハッシュメニュー応用編」-「最小完全ハッシュ関数を作ろう」に
swiftでチャレンジしたコードです。
https://paiza.jp/works/mondai/hash_advanced/hash_advanced__minimum_complete

作成環境
Ubuntu 22.04.5 LTS
Swift version 5.4 (swift-5.4-RELEASE) Target: x86_64-unknown-linux-gnu
*/
import Foundation

// ハッシュ関数X
// parameters
//   s: 文字列
// return sのハッシュ値
func hashFunction(_ s: String) -> Int {
    let string = "abcdef"
    var total = 0
    let length = s.count
    for (i, char) in zip(s.indices, s) {
        if let index = string.firstIndex(of: char)?.utf16Offset(in: string) {
            let iInt = i.utf16Offset(in: s)
            total += index * Int(pow(6.0, Double(length - iInt - 1)))
        }
    }
    return total
}

// 最小完全ハッシュ関数を作る
func main() {
    // 文字列 s が与えられます。
    let s = readLine()!.trimmingCharacters(in: .whitespacesAndNewlines)
    // ハッシュ値を求める
    let hashValue = hashFunction(s)
    // 結果出力
    print(hashValue)
}
// エントリーポイント
main()
